File: samples/src/main/java/ru/vafeen/samples/sample2/java/CollectionsMapper.java
package ru.vafeen.samples.sample2.java;

import static ru.vafeen.samples.sample2.java.CollectionsMapperCastCastleKt.mapACastCastle;
import static ru.vafeen.samples.sample2.java.CollectionsMapperCastCastleKt.mapBCastCastle;
import static ru.vafeen.samples.sample2.java.CollectionsMapperCastCastleKt.mapLevel1ACastCastle;
import static ru.vafeen.samples.sample2.java.CollectionsMapperCastCastleKt.mapLevel1BCastCastle;

import ru.vafeen.castcastle.annotations.CastCastleMapper;

@CastCastleMapper
public interface CollectionsMapper {

    @CastCastleMapper
    default B mapA(A a) {
        return mapACastCastle(this, a);
    }

    @CastCastleMapper
    default A mapB(B b) {
        return mapBCastCastle(this, b);
    }

    default int string(String string) {
        return Integer.parseInt(string);
    }

    default String mapInt(int i) {
        return String.valueOf(i);
    }

    @CastCastleMapper
    default InnerLevel1B mapLevel1A(InnerLevel1A inner1Level1) {
        return mapLevel1ACastCastle(this, inner1Level1);
    }

    @CastCastleMapper
    default InnerLevel1A mapLevel1B(InnerLevel1B inner1Level1) {
        return mapLevel1BCastCastle(this, inner1Level1);
    }

    //    @CastCastleMapper
    //    default List<InnerLevel1B> mapLevel1A(List<InnerLevel1A> inner1Level1) {
    //        return mapLevel1AListCastCastle(inner1Level1);
    //    }
    //
    //    @CastCastleMapper
    //    default List<InnerLevel1A> mapLevel1B(List<InnerLevel1B> inner1Level1) {
    //        return mapLevel1BListCastCastle(inner1Level1);
    //    }
    //
    //    @CastCastleMapper
    //    default InnerLevel2B mapLevel2A(InnerLevel2A innerLevel2A) {
    //        return mapLevel2ACastCastle(innerLevel2A);
    //    }
    //
    //    @CastCastleMapper
    //    default InnerLevel2A mapLevel2B(InnerLevel2B innerLevel2A) {
    //        return mapLevel2BCastCastle(innerLevel2A);
    //    }
    //
    //    @CastCastleMapper
    //    default List<InnerLevel2B> mapLevel2A(List<InnerLevel2A> innerLevel2A) {
    //        return mapLevel2AListCastCastle(innerLevel2A);
    //    }
    //
    //    @CastCastleMapper
    //    default List<InnerLevel2A> mapLevel2B(List<InnerLevel2B> innerLevel2A) {
    //        return mapLevel2BListCastCastle(innerLevel2A);
    //    }
}

File: samples/src/main/java/ru/vafeen/samples/sample2/kotlin/CollectionsMapper.kt
package ru.vafeen.samples.sample2.kotlin

import ru.vafeen.castcastle.annotations.CastCastleMapper

data class A(val inner1Level1: List<InnerLevel1A>, val inner2Level1: InnerLevel1A)
data class B(val inner1Level1: List<InnerLevel1B>, val inner2Level1: InnerLevel1B)
data class InnerLevel1A(val inner1Level2: List<InnerLevel2A>, val inner2Level2: InnerLevel2A)
data class InnerLevel1B(val inner1Level2: List<InnerLevel2B>, val inner2Level2: InnerLevel2B)
data class InnerLevel2A(val x: String, val y: Int)
data class InnerLevel2B(val x: Int, val y: String)

@CastCastleMapper
internal interface CollectionsMapper {
    @CastCastleMapper
    fun mapA(a: A): B = mapACastCastle(a)

    @CastCastleMapper
    fun mapB(b: B): A = mapBCastCastle(b)

    fun string(string: String): Int = string.toInt()
    fun int(int: Int): String = "$int"

    @CastCastleMapper
    fun mapLevel1A(inner1Level1: InnerLevel1A): InnerLevel1B = mapLevel1ACastCastle(inner1Level1)

    @CastCastleMapper
    fun mapLevel1B(inner1Level1: InnerLevel1B): InnerLevel1A = mapLevel1BCastCastle(inner1Level1)

    //    fun mapLevel1A(inner1Level1: List<InnerLevel1A>): List<InnerLevel1B>
    //    fun mapLevel1B(inner1Level1: List<InnerLevel1B>): List<InnerLevel1A>
    //    fun mapLevel2A(innerLevel2A: InnerLevel2A): InnerLevel2B
    //    fun mapLevel2B(innerLevel2A: InnerLevel2B): InnerLevel2A
    //    fun mapLevel2A(innerLevel2A: List<InnerLevel2A>): List<InnerLevel2B>
    //    fun mapLevel2B(innerLevel2A: List<InnerLevel2B>): List<InnerLevel2A>
}